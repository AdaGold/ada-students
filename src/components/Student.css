/*Student.css*/
.student .absent {
  color: red;
}

.student .present {
  color: green;
}
