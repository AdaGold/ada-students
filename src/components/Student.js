import React from 'react';
import PropTypes from 'prop-types';

import './Student.css';

const Student = (props) => {

  // Event callback functions
  const onButtonClick = () => {
    const updatedStudent = {
      fullName: props.fullName,
      birthday: props.birthday,
      email: props.email,
      present: !props.present,
      id: props.id,
    }
    props.onUpdateStudent(updatedStudent);
  }

  const onFullNameInputChange = (event) => {
    const fullName = event.target.value;
    props.onUpdateStudent({
      fullName,
      birthday: props.birthday,
      email: props.email,
      present: props.present,
      id: props.id,
    });
  };

  // Component functions always return JSX
  return (
    <div className="student">
      <h3 className={props.present ? 'present' : 'absent'}>{props.fullName}</h3>
      <input value={props.fullName} onChange={onFullNameInputChange} />
      <ul>
        <li>Class: C13</li>
        <li>Birthday: {props.birthday}</li>
        <li>Email: {props.email}</li>
      </ul>
      <button onClick={onButtonClick}>
        Mark {props.present ? 'Absent' : 'Present'}
      </button>
    </div>
  );
};

Student.propTypes = {
  fullName: PropTypes.string.isRequired,
  email: PropTypes.string,
  birthday: PropTypes.string,
  onUpdateStudent: PropTypes.func.isRequired,
};

Student.defaultProps = {
  birthday: 'no birthdate on file',
}

export default Student;
