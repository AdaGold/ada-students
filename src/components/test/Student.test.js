// src/components/test/Student.test.js
import React from 'react';
import { render, cleanup } from '@testing-library/react'
import Student from '../Student';

describe('Student', () => {

  afterEach(cleanup);

  test('that it matches the existing snapshot', () => {
    // Arrange-Act
    const { asFragment } = render(
      <Student
        fullName="Ada Lovelace"
        email="Ada@adadev.org"
        birthday="10 December 1815"
        onUpdateStudent={() => { }}
      />
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
    cleanup();
  });

  test('The fullName field renders properly', () => {

    const { getByText } = render(<Student
      fullName="Ada Lovelace"
      email="Ada@adadev.org"
      birthday="10 December 1815"
      present={true}
      onUpdateStudent={() => { }}
    />);

    const element = getByText(/Ada Lovelace/i);
    expect(element).toBeInTheDocument();
  });

  test('The Student\'s button renders properly when they are present', () => {

    const { getByText } = render(<Student
      fullName="Ada Lovelace"
      email="Ada@adadev.org"
      birthday="10 December 1815"
      present={true}
      onUpdateStudent={() => { }}
    />);

    // because the student is present
    const element = getByText(/Mark Absent/i);
    expect(element).toBeInTheDocument();
  });

  test('The Student\'s button renders properly when they are absent', () => {

    const { getByText } = render(<Student
      fullName="Ada Lovelace"
      email="Ada@adadev.org"
      birthday="10 December 1815"
      present={false}
      onUpdateStudent={() => { }}
    />);

    // because the student is present
    const element = getByText(/Mark Present/i);
    expect(element).toBeInTheDocument();
  });

  test('The "onUpdateStudent" prop function is called when the `Mark Absent` button is clicked on', () => {

    // Arrange
    // Create a mock callback function
    const markAbsent = jest.fn();

    // Render a Student
    //   note the markAbsent callback function
    const container = render(<Student
      fullName="Ada Lovelace"
      email="Ada@adadev.org"
      birthday="10 December 1815"
      present={true}
      onUpdateStudent={markAbsent}
    />);

    // Act
    // Find the Mark Absent Button
    const button = container.getByText(/Mark Absent/i);

    // Trigger a click event
    button.click();

    // Assert
    // Verify that the callback function was called
    expect(markAbsent).toHaveBeenCalled();
  });
});
