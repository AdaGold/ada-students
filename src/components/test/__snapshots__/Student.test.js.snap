// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student that it matches the existing snapshot 1`] = `
<DocumentFragment>
  <div
    class="student"
  >
    <h3
      class="absent"
    >
      Ada Lovelace
    </h3>
    <input
      value="Ada Lovelace"
    />
    <ul>
      <li>
        Class: C13
      </li>
      <li>
        Birthday: 10 December 1815
      </li>
      <li>
        Email: Ada@adadev.org
      </li>
    </ul>
    <button>
      Mark Present
    </button>
  </div>
</DocumentFragment>
`;
